[MULTICLIENT003] Add new analyzer and fix for usage of `ItemFlags` in `switch` statements.

Ijwu opened the pull request:
Closes: #2 

Adds a new analyzer warning (MULTICLIENT003) and associated fix for the usage of `ItemFlags` within a `switch` statement. Since `switch` statements perform direct comparison, they run afoul of the standard set in MULTICLIENT002.

It was decided to add a new analyzer after discussion in the AP discord.

See the updated README for info on the new analyzer and fix, then check out the unit tests for how it works.

Commit:
Update unit testing packages

Commit:
Add new diagnostic for ItemFlags used as the target label/expression in a switch statement.

Commit:
Rename old item flags tests and add new tests for itemflags in switch statements.

Commit:
Remove duplicated block from README

Commit:
Remove tabs from README.

Commit:
Resolve minor quibbles with the README

Commit:
Remove note on ItemFlags.Filler from MULTICLIENT003 docs

Commit:
Add fix for MULTICLIENT003

Commit:
Remove unused variable.

BadMagic100 commented on Archipelago.MultiClient.Net.Analyzers/Archipelago.MultiClient.Net.Analyzers.Test/ItemFlagsInSwitchTest.cs:
```diff
@@ -0,0 +1,149 @@
+namespace Archipelago.MultiClient.Net.Analyzers.Test
+{
+    [TestClass]
+    public class ItemFlagsInSwitchTest
```
I think a test to confirm the correct generation of else-if's would be nice here

Ijwu replied:
Good call out, thanks. Added this in.

BadMagic100 commented on Archipelago.MultiClient.Net.Analyzers/Archipelago.MultiClient.Net.Analyzers/AnalyzerReleases.Unshipped.md:
```diff
@@ -1,3 +1,6 @@
 ﻿; Unshipped analyzer release
 ; https://github.com/dotnet/roslyn-analyzers/blob/main/src/Microsoft.CodeAnalysis.Analyzers/ReleaseTrackingAnalyzers.Help.md
 
+Rule ID | Category | Severity | Notes
```
```suggestion
### New Rules

Rule ID | Category | Severity | Notes
```

BadMagic100 commented on Archipelago.MultiClient.Net.Analyzers/Archipelago.MultiClient.Net.Analyzers/Fixes/ItemFlagsInSwitchFixes.cs:
```diff
@@ -0,0 +1,137 @@
+
+            context.RegisterCodeFix(
+                    CodeAction.Create(
+                        title: "Convert ItemFlags switch to if/else",
```
```suggestion
                        title: "Convert to if/else",
```

This seems more stylistically correct

BadMagic100 commented on Archipelago.MultiClient.Net.Analyzers/Archipelago.MultiClient.Net.Analyzers/Fixes/ItemFlagsInSwitchFixes.cs:
```diff
@@ -0,0 +1,137 @@
+            foreach (var node in switchSections.Reverse())
+            {
+                SwitchSectionSyntax switchSection = (SwitchSectionSyntax)node;
+                List<ExpressionSyntax> cases = switchSection.Labels.OfType<CaseSwitchLabelSyntax>().Select(x => x.Value).ToList();
```
I think we might have to watch out for pattern matching here, for most kinds of patterns it would be invalid syntax anyway but e.g. the user could potentially get away with something like `case var f when f.HasFlag(ItemFlags.Advancement)` which is not a violation of the rule but would result in this fix producing bad syntax

BadMagic100 replied:
Cases with bitwise | could also be a concern (assuming they are legal) in that maybe in that case the expected comparison actually should be ==

Ijwu replied:
Verified this is an issue in a test. Will see if I can find a solution with some finagling.

BadMagic100 reviewed (changes requested):
General asks:
* Please use nullability annotations where appropriate (see compiler warnings)
* Please don't use `var` except for destructuring assignments (i don't think there were any but I feel like the typing on those gets ugly otherwise so that's my exception to the rule)

I also am concerned about switch expressions in the sense they are not covered here, that could be a separate issue/PR though. If included here I think it's fine for it to be the same rule ID/analyzer, in theory it seems like it's possible to have multiple fix providers on the same rule as long as they are implemented in a way that they are aware they might not be able to fix everything if the fixes needed to be separated

Ijwu committed:
Update unshipped analyzer releases file

Co-authored-by: BadMagic100 <dempsey.sean@outlook.com>

Ijwu committed:
Change itemflags in switch code fix to have a more concise title

Co-authored-by: BadMagic100 <dempsey.sean@outlook.com>

Commit:
Touch up usage of `var` and nullability annotations.

I don't like to use the null-forgiving operator generally, but I didn't have a better idea this time.

Ijwu commented:
@BadMagic100 and I discussed a new way to go with this. I will work on that over time and will un-draft the PR when it is ready.

Commit:
[WIP] Prepare to switch MULTICLIENT003 to case statement analysis

Commit:
Re-implement MULTICLIENT003 diagnostic to target case statements directly

Commit:
[WIP] Prepare unit tests for new fix implementation for MULTICLIENT003

BadMagic100 commented on Archipelago.MultiClient.Net.Analyzers/Archipelago.MultiClient.Net.Analyzers.Test/ItemFlagComparisonsTest.cs:
```diff
@@ -159,11 +246,18 @@ namespace MyClient
+            {
+                {|#0:case ItemFlags.Advancement:|}
+                    return true;
+                {|#1:case ItemFlags.None:|}
```
Exact comparisons to ItemFlags.None should be allowed

Ijwu replied:
Fixed, comparisons to `ItemFlags.None` now do not trigger the diagnostic. Unit tests have been revised for this.

BadMagic100 commented on Archipelago.MultiClient.Net.Analyzers/Archipelago.MultiClient.Net.Analyzers/Analyzers/ItemFlagsComparisonInSwitchCaseDiagnostics.cs:
```diff
@@ -4,25 +4,26 @@
-    // TODO - how much of this applies to switch expressions or pattern matching switch statements?
     [DiagnosticAnalyzer(LanguageNames.CSharp)]
-    public class ItemFlagsInSwitchDiagnostics : DiagnosticAnalyzer
+    public class ItemFlagsComparisonInSwitchCaseDiagnostics : DiagnosticAnalyzer
```
Class name should match the file name. I also think this name is a bit clunky though the only alternatives coming to mind are worse

Ijwu replied:
It does match the file name, doesn't it? I don't think I'm missing something there. The name is indeed clunky, I'll see if I can come up with something better but I probably won't think about it too hard.

BadMagic100 replied:
For some reason I thought it didn't, I must have misread the diff

BadMagic100 commented on Archipelago.MultiClient.Net.Analyzers/Archipelago.MultiClient.Net.Analyzers/Analyzers/ItemFlagsComparisonInSwitchCaseDiagnostics.cs:
```diff
@@ -35,15 +36,23 @@ public override void Initialize(AnalysisContext context)
-                return;
+                foreach (CaseSwitchLabelSyntax label in section.Labels.OfType<CaseSwitchLabelSyntax>())
+                {
+                    if (label.Value is MemberAccessExpressionSyntax identifier)
```
There are potentially ways to slip through this with badly behaved (non hasflags) pattern matches, I'm not sure how reasonable or common they are though. It might be worth it to flag any pattern matching that isn't the kind we know is good (obviously we cannot offer a fix but at least we can raise the signal)

BadMagic100 replied:
After thinking on this and looking at the current state of the analyzer now that it accommodates switch expressions, I don't want it to get too complicated and we already have some signals that overly aggressive warnings on other analyzers are annoying. So I will leave it out.

Commit:
[WIP] Allow for direct comparisons to `ItemFlags.None`

Commit:
[WIP] Update unit tests away from using `ItemFlags.None` as a problematic case

Commit:
[WIP] First attempt at updated fix provider for MULTICLIENT003

BadMagic100 commented on Archipelago.MultiClient.Net.Analyzers/Archipelago.MultiClient.Net.Analyzers/Analyzers/ItemFlagsComparisonInSwitchCaseDiagnostics.cs:
```diff
@@ -47,12 +47,15 @@ private void AnalyzeSwitch(SyntaxNodeAnalysisContext context)
                         {
-                            context.ReportDiagnostic(Diagnostic.Create(NoItemFlagsComparisonsInSwitchCaseConstants, label.GetLocation()));
+                            // Check if the identifier is ItemFlags.None
+                            if (identifier.Name.ToString() != "None")
```
I think the better way to do this is by checking the actual integer value of the enum is 0, I do this in the if statement version for reference. That would make it more resilient to renames

Ijwu replied:
I was able to make this happen with `SemanticModel.GetConstantValue()`, so that's nice.

Ijwu committed:
Remove fixall provider for MULTICLIENT003 fixer

Ijwu committed:
Update README and unshipped releases on MULTICLIENT003

BadMagic100 commented on Archipelago.MultiClient.Net.Analyzers/Archipelago.MultiClient.Net.Analyzers.Test/ItemFlagsTest.cs:
```diff
@@ -1,267 +0,0 @@
-﻿using Microsoft.CodeAnalysis.Testing;
```
This file (testing diagnostics and fixes for == comparisons) should be kept, I assume you meant to rename it (my preference) or merge it with the one for switches

BadMagic100 commented on Archipelago.MultiClient.Net.Analyzers/Archipelago.MultiClient.Net.Analyzers/Analyzers/ItemFlagsComparisonInSwitchCaseDiagnostics.cs:
```diff
@@ -0,0 +1,61 @@
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+
+            context.RegisterSyntaxNodeAction(AnalyzeSwitch, SyntaxKind.SwitchStatement);
```
Did we also try to apply this to switch expressions yet?

Ijwu replied:
I have left switch expressions as "out of scope for now". Partially to keep things simple and mostly because I'm lazy.

BadMagic100 commented on Archipelago.MultiClient.Net.Analyzers/Archipelago.MultiClient.Net.Analyzers/Fixes/ItemFlagsInSwitchFixes.cs:
```diff
@@ -0,0 +1,129 @@
+
+        private static string GetPatternMatchingVariableName(DocumentEditor editor, SwitchStatementSyntax switchStatement)
+        {
+            var analysis = editor.SemanticModel.AnalyzeDataFlow(switchStatement);
```
You might also try just analyzing the specific case to try to deal with the scoping stuff we were discussing

BadMagic100 commented on Archipelago.MultiClient.Net.Analyzers/Archipelago.MultiClient.Net.Analyzers/Fixes/ItemFlagsInSwitchFixes.cs:
```diff
@@ -0,0 +1,129 @@
+                return "f";
+            }
+            
+            var hasCollision = analysis.WrittenInside.Any(x => x.Name.StartsWith("f"));
```
In keeping with the above comment, I think you may be looking for CapturedInside, CapturedOutside, or DataFlowsIn (to see what all variables are available to the scope). I think we also need a test which has an unrelated `f` defined outside the switch to make sure we avoid that scoping collision

Ijwu replied:
`WrittenInside` was one of the only ones actually showing the `f` variable in it. I hadn't considered an unrelated `f`, though, which will likely change things.

BadMagic100 commented on Archipelago.MultiClient.Net.Analyzers/Archipelago.MultiClient.Net.Analyzers/Fixes/ItemFlagsInSwitchFixes.cs:
```diff
@@ -0,0 +1,129 @@
+                    SyntaxFactory.MemberAccessExpression(
+                        SyntaxKind.SimpleMemberAccessExpression,
+                        SyntaxFactory.IdentifierName(newVarName),
+                        SyntaxFactory.IdentifierName("HasFlag")
```
Just calling out that net35 does not have HasFlag - idk if/how we can detect the target runtime and adjust accordingly

BadMagic100 replied:
This is possible to detect but I will leave it out of scope for now since this problem exists elsewhere. I think #1 covers this actually

BadMagic100 commented on Archipelago.MultiClient.Net.Analyzers/Archipelago.MultiClient.Net.Analyzers/Fixes/ItemFlagsInSwitchFixes.cs:
```diff
@@ -0,0 +1,129 @@
+        private CasePatternSwitchLabelSyntax GeneratePatternMatchingCaseLabel(CaseSwitchLabelSyntax caseLabel, ExpressionSyntax switchExpression, string newVarName)
+        {
+            // Create the pattern matching statement: "case var f when f.HasFlag(ItemFlags.Advancement):"
+            var variableDesignation = SyntaxFactory.SingleVariableDesignation(SyntaxFactory.Identifier(newVarName));
```
general reminder I prefer explicit typing rather than `var` (ironically I don't think we can avoid this for the generated pattern matching clauses)

Ijwu replied:
Whoops. Usually I plop `var` in then go back and make them explicit. I did not do the second step. Will fix that up in next commit.

BadMagic100 commented on Archipelago.MultiClient.Net.Analyzers/Archipelago.MultiClient.Net.Analyzers.Test/ItemFlagComparisonsTest.cs:
```diff
@@ -0,0 +1,427 @@
+    
+
+        [TestMethod]
+        public async Task VerifyItemFlagsInSwitchExpressionYieldsNoDiagnostic()
```
assuming that we were to keep switch expressions out of this, this is not a good test as none of the cases here would violate the rule normally.

Ijwu replied:
That test may not be named in the best way, but its intent is to ensure that no diagnostic is issued for a compliant case.

BadMagic100 replied:
I guess I'm saying if we are leaving switch expressions intentionally out of scope we should also have a test which asserts no diagnostic on a switch expression which would be a violation (with a comment that it should be changed if/when we do support switch expressions)

Ijwu replied:
I've updated the test case to include a problematic match in the switch expression, while still asserting it yields no diagnostic for switch expressions. 

Sorry for not tackling switch expressions, but I'm nearing completion on this and it'd take me another 2 months to gather the spirit bomb energy to do more. 😓

BadMagic100 replied:
I found the spirit bomb energy today

Ijwu committed:
MULTICLIENT003 diagnostic now checks enum value against 0 to determine `ItemFlags.None`

Also updated help link

Ijwu committed:
Use explicit typing in MULTICLIENT003 fix

Ijwu committed:
[MULTICLIENT003] Add problematic case to test verifying switch expressions yield no diagnostic

Ijwu committed:
[MULTICLIENT003] Add new tests containing unrelated variable named "f"

Ijwu committed:
[MULTICLIENT003] Handle unrelated `f` variables in fix.

Also removed unused parameter from a method.

BadMagic100 committed:
Cleaner unique name generation

BadMagic100 committed:
Rename ItemFlagSwitchTest in prep to restore old comparison tests

BadMagic100 committed:
Recover MULTICLIENT002 tests

BadMagic100 committed:
Apply MULTICLIENT003 and fixes to switch expressions, change to explicitly typed patterns

BadMagic100 committed:
Update multiclient003 fixed code in readme
